### PR TITLE
Fix WASI rename issue: #2297

### LINKED ISSUE #2297: error: unable to rename temporary
### Describe the bug

Ubuntu 21.04
wasmer 1.0.2 | x86_64

Trying to run `clang.wasm`, getting `error: unable to rename temporary 'source/hello/hello-1df6f5ea.o.tmp' to output file 'source/hello/hello.o': 'I/O error'`.
It seemed to work when the source/destination were in the root, but fails in subfolders.

https://github.com/wasm3/wasm3-self-compiling/blob/7a251d9483cdd4f559e7cf4af36a6468af100ea4/Makefile#L21-L26

### Steps to reproduce

```sh
git clone https://github.com/wasm3/wasm3-self-compiling.git

cd wasm3-self-compiling

wasmer run --mapdir=/:. --mapdir=./:. ./wasm/clang.wasm -- -cc1 -triple wasm32-unknown-wasi -isysroot /sys -internal-isystem /sys/include -emit-obj -o source/hello/hello.o source/hello/hello.c -O3
```

### Expected behavior

Compilation succeeds.

### Actual behavior

`error: unable to rename temporary 'source/hello/hello-1df6f5ea.o.tmp' to output file 'source/hello/hello.o': 'I/O error'`

### PR DESCRIPTION
**Pull request by chenyukang**
# Description
Fix issue #2297.

The old code push full path to the parent directory, if we run wasmer as below: 

```bash
wasmer  run --mapdir=/:. --mapdir=./:.  ./wasm/clang.wasm -- -cc1 -triple wasm32-unknown-wasi -isysroot /sys -internal-isystem /sys/include -emit-obj -o ./source/hello/
hello.o ./source/hello/hello.c
```
The `out_path` become to : `./source/hello/source/hello/hello.o`, it will trigger an I/O error.

close #2297

### CODE REVIEW

**Comment by syrusakbary** [comment]
Good fix! I'll wait for @MarkMcCaskey review as he has more insight on the WASI internals :)

**Comment by chenyukang** [comment]
Seems we don't have 'make wasitests' now.
https://github.com/wasmerio/wasmer/tree/master/tests/wasi-wast/wasi/test_fs

File: lib/wasi/src/syscalls/mod.rs [hunk 1]
```diff
@@ -2065,7 +2065,7 @@ pub fn path_rename(
                 return __WASI_EEXIST;
             }
             let mut out_path = path.clone();
-            out_path.push(target_path);
+            out_path.push(target_path.file_name().unwrap());
```
[the remark this answers is not in the document]
Comment (chenyukang) [reply lib/wasi/src/syscalls/mod.rs:2068]: Yes! `target_entry_name` is the same value and also verified one.

File: tests/wasi-wast/wasi/tests/path_rename.rs [hunk 2]
```diff
@@ -73,3 +73,47 @@ fn main() {
     println!("{}", test_str);
     std::fs::remove_file(file_to_rename_to).unwrap();
 }
+
+fn run_with_sub_dir() {
```
[the remark this answers is not in the document]
Comment (chenyukang) [reply tests/wasi-wast/wasi/tests/path_rename.rs]: I made another PR for wasi-tests: https://github.com/wasmerio/wasi-tests/pull/1

[the remark this answers is not in the document]
Comment (chenyukang) [reply tests/wasi-wast/wasi/tests/path_rename.rs]: @MarkMcCaskey Please have a review on the tests :)

**Comment by syrusakbary** [comment]
Can you update the wasi-tests subtree to latest master (with your fix added)?

**Review by syrusakbary — APPROVED** [review]

**Pull request merged**